### PR TITLE
Add swiftpm manifest/lockfile parser

### PR DESCRIPTION
**Pull request by jarrydlee**
We're adding support for the swift language in the supply chain product. To do this we need to add parsers for the manifest and lockfile. We are currently only adding support for the `Swift Package Manager`.

requires https://github.com/semgrep/semgrep-interfaces/pull/217

### CODE REVIEW

**Comment by jarrydlee** [comment]
Waiting on 1.59 release for `semgrep-interfaces` so that we can pull in the correct submodule with `PackageSwift` type

File: cli/src/semdep/parsers/swiftpm.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+from semgrep.semgrep_interfaces.semgrep_output_v1 import SwiftPM
+from semgrep.semgrep_interfaces.semgrep_output_v1 import Transitive
+from semgrep.semgrep_interfaces.semgrep_output_v1 import Transitivity
+from semgrep.verbose_logging import getLogger
+
+
+logger = getLogger(__name__)
+
+
+package_name = string('"') >> upto('"').bind(lambda url: success(url))
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py]: Might actually be easier to read as as a regex

File: cli/src/semdep/parsers/swiftpm.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                transitivity=get_transitivity(package_name, direct_deps),
+                line_number=version.line_number,
+            )
+        )
+
+    return result
+
+
+def extract_package_name(package_uri: str) -> str:
+    return package_uri.split("/")[-1].replace(".git", "")
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py]: This makes me nervous about package uris with unexpected formats. I think we should explicitly handle and exception here

File: cli/src/semdep/parsers/swiftpm.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    )
+
+    if not parsed_lockfile or not parsed_manifest:
+        return [], errors
+
+    direct_deps = parse_manifest_deps(parsed_manifest)
+    lockfile_json = parsed_lockfile.as_dict()
+    lockfile_version = lockfile_json.get("version")
+    if lockfile_version is None:
+        logger.info("no version in lockfile %s", lockfile_path)
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py:223]: Maybe this should be something added to `errors`?

File: cli/src/semdep/parsers/swiftpm.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    << any_char.until(string("\n"))
+)
+
+comment = string("\n").optional() >> regex(r" *//([^\n]*)", flags=0, group=1)
+
+package_swift_parser = (
+    any_char.until(regex(r"dependencies\s*:\s*\[.*?\n"))
+    >> (comment | package_section)
+    .sep_by(string("\n"))
+    .map(lambda xs: [x for x in xs if isinstance(x, tuple)])
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py]: might be easier to read if this was separated out and given it's own name

File: cli/src/semdep/parsers/swiftpm.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    << any_char.many()
+)
+
+
+def get_transitivity(package_name: str, direct_deps: Set[str]) -> Transitivity:
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py]: this function already exists in utils, it's called `transitivity`.

File: cli/src/semdep/parsers/swiftpm.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    if not lockfile_version.as_int():
+        return [], errors
+
+    all_deps = []
+    if lockfile_version == 1:
+        all_deps = parse_swiftpm_v1(lockfile_json, direct_deps)
+    elif lockfile_version == 2:
+        all_deps = parse_swiftpm_v2(lockfile_json, direct_deps)
+
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py:250]: What if it's neither? Should have an error case

**Review by mmcqd — APPROVED** [review]
Overall looks great!

**Comment by jarrydlee** [comment]
Still working on last couple test/build failures

File: cli/src/semdep/parsers/swiftpm.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+# url: "https://example.com/example-package.git"
+url_block = regex(r'url:\s*"') >> git_url << string('"')
+
+separator_block = regex(r"\s*,\s*")
+
+# from: "1.2.3"
+from_block = regex(r'from:\s*".*?"')
+
+# "1.2.3"..<"1.2.6"
```
Comment (mmcqd) [inline cli/src/semdep/parsers/swiftpm.py:55]: Did you get these from somewhere? If so you should cite it
"these" meaning the various regexes and examples

**Review by mmcqd — APPROVED** [review]
hell yeah brother

**Pull request merged**